### PR TITLE
Added auto-select midi device feature in response to issue #65

### PR DESCRIPTION
**Pull request by thealienthing**
I added an auto-connect feature to the preferences window in February. It was just an on/off button so if both buttons for auto-connecting in a new window and auto-connecting in a current window were selected, it combined both behaviors. Most recent commits in June changed the UI so that the preferences were chosen using a radio button.

This is my first real pull request to an open source project. Please let me know if I need to change anything. I'm happy to do anything necessary in order to be able to participate in this project. Thank you!

### CODE REVIEW

**Comment by thealienthing** [comment]
Added the uninstall instructions to the documentation.

**Comment by krevis** [comment]
Hi, thank you for the PR. it has been an exceptionally busy few months for me (iOS 13...) and I'm trying to catch up. Taking a look.

I think the uninstall instructions change (#68) would ideally be a separate PR.

File: Applications/MIDIMonitor/English.lproj/docs.htmld/index.html [hunk 1]
```diff
@@ -35,6 +35,7 @@
             <li><a href="#howto">How to use it</a></li>
             <li><a href="#preferences">Preferences</a></li>
             <li><a href="#advanced">Advanced</a></li>
+            <li><a href="#uninstall">Uninstall</a></li>
```
Comment (krevis) [inline Applications/MIDIMonitor/English.lproj/docs.htmld/index.html]: Let's please do this in a separate PR, since it's unrelated to the other changes.

File: Applications/MIDIMonitor/English.lproj/docs.htmld/index.html [hunk 2]
```diff
@@ -311,6 +312,28 @@ <h3>Rescan MIDI</h3>
         </div>
       </div>
 
+      <div class="page-header">
+        <h1 id="uninstall">Uninstall</h1>
+      </div>
+
+      <div class="row">
+        <div class="col-md-7">
+          <h3>How to uninstall Midi Monitor</h3>
```
Comment (krevis) [inline Applications/MIDIMonitor/English.lproj/docs.htmld/index.html]: MIDI not Midi

File: Applications/MIDIMonitor/SMMAppController.h [hunk 3]
```diff
@@ -19,4 +19,7 @@
 @end
 
 // Preference keys
-extern NSString* const SMMOpenWindowsForNewSourcesPreferenceKey;
+extern NSString* const SMMAutoConnectNewSourcesPreferenceKey;
+//extern NSString* const SMMOpenWindowsForNewSourcesPreferenceKey;
```
Comment (krevis) [inline Applications/MIDIMonitor/SMMAppController.h]: Please delete this, there is no need to leave commented-out code around.

File: Applications/MIDIMonitor/SMMAppController.m [hunk 4]
```diff
@@ -332,9 +331,21 @@ - (void)failedToConnectToSpyClient
 
 - (void)sourceEndpointsAppeared:(NSNotification *)notification
 {
-    if ([[NSUserDefaults standardUserDefaults] boolForKey:SMMOpenWindowsForNewSourcesPreferenceKey]) {
+    NSInteger option = [[NSUserDefaults standardUserDefaults] integerForKey:SMMAutoConnectNewSourcesPreferenceKey];
```
Comment (krevis) [inline Applications/MIDIMonitor/SMMAppController.m]: It would be nice if there were a migration path from the old default to the new one. As it is now, you're completely ignoring the old preference, which is a bit rude to our users.

You could, for instance, add code to `applicationWillFinishLaunching:` that reads the old default and sets the corresponding value for the new default. (There is no need to remove the old default afterwards -- it's nice to leave it around in case someone runs an older version of the app.)

File: Applications/MIDIMonitor/SMMAppController.m [hunk 5]
```diff
@@ -332,9 +331,21 @@ - (void)failedToConnectToSpyClient
...
+    
+    if (option == 1) {
```
Comment (krevis) [inline Applications/MIDIMonitor/SMMAppController.m]: Please make an enum for these values somewhere, instead of using magic numbers 1 and 2.

File: Applications/MIDIMonitor/SMMAppController.m [hunk 6]
```diff
@@ -332,9 +331,21 @@ - (void)failedToConnectToSpyClient
...
         NSArray *endpoints = [[notification userInfo] objectForKey:SMMIDIObjectsThatAppeared];
-
+        
+        if (!self.newlyAppearedSources) {
+            self.newlyAppearedSources = [NSMutableSet set];
+            [self performSelector:@selector(autoConnectToNewlyAppearedSources) withObject:nil afterDelay:0.1 inModes:@[NSDefaultRunLoopMode]];
```
Comment (krevis) [inline Applications/MIDIMonitor/SMMAppController.m]: Since the code for cases 1 and 2 are exactly the same, aside from the `@selector`, it would be better to avoid repeating the code. Move the check for option == 1 or option == 2 down here, and use it to choose a different selector.

File: Applications/MIDIMonitor/SMMCombinationInputStream.m [hunk 7]
```diff
@@ -139,7 +139,7 @@ - (NSSet *)selectedInputSources
 
     return inputSources;
 }
-
+//WORK HERE
```
Comment (krevis) [inline Applications/MIDIMonitor/SMMCombinationInputStream.m]: Please remove things like this before putting up a PR.

File: Applications/MIDIMonitor/SMMPreferencesWindowController.m [hunk 8]
```diff
@@ -167,6 +163,12 @@ - (IBAction)changeExpertMode:(id)sender
     [self sendDisplayPreferenceChangedNotification];
 }
 
+- (IBAction)changeNewSourcesRadio:(id)sender
+{
+    [[NSUserDefaults standardUserDefaults] setInteger:[[sender selectedCell] tag] forKey:SMMAutoConnectNewSourcesPreferenceKey];
+    [self sendDisplayPreferenceChangedNotification];
```
Comment (krevis) [inline Applications/MIDIMonitor/SMMPreferencesWindowController.m]: This isn't a "display preference" (it doesn't affect how messages are displayed) so you don't need to post this notification.

File: Applications/MIDIMonitor/SMMAppController.m [hunk 9]
```diff
@@ -357,4 +368,15 @@ - (void)openWindowForNewlyAppearedSources
     self.newlyAppearedSources = nil;
 }
 
+- (void)autoConnectToNewlyAppearedSources
+{
+    NSDocumentController *dc = [NSDocumentController sharedDocumentController];
+    SMMDocument *document = [dc currentDocument];
+    [document setSelectedInputSources:self.newlyAppearedSources];
+    SMMMonitorWindowController *wc = document.windowControllers.firstObject;
```
Comment (krevis) [inline Applications/MIDIMonitor/SMMAppController.m]: If there are multiple windows, you are doing this for only one of them. Shouldn't it change all of them?

File: Applications/MIDIMonitor/English.lproj/docs.htmld/index.html (hunk 1 above)
Comment (thealienthing) [reply Applications/MIDIMonitor/English.lproj/docs.htmld/index.html]: Okay. Can I undo this just using a revert to the last commit? Or is there a better way?

File: Applications/MIDIMonitor/SMMAppController.h (hunk 3 above)
Comment (krevis) [reply Applications/MIDIMonitor/SMMAppController.h]: (Although you probably want to keep the old preference key around for migration purposes.)

File: Applications/MIDIMonitor/English.lproj/docs.htmld/index.html (hunk 1 above)
Comment (krevis) [reply Applications/MIDIMonitor/English.lproj/docs.htmld/index.html]: Yes, I think that should work.

File: Applications/MIDIMonitor/SMMAppController.m (hunk 9 above)
Comment (thealienthing) [reply Applications/MIDIMonitor/SMMAppController.m]: My thought was to autoconnect to new devices in the current window only. But I can change it so that it works with all windows.

File: Applications/MIDIMonitor/SMMAppController.h (hunk 3 above)
Comment (thealienthing) [reply Applications/MIDIMonitor/SMMAppController.h]: Would you prefer I just delete the one comment but leave the old OpenWindowsForNewSourcesPreferenceKey?

File: Applications/MIDIMonitor/SMMCombinationInputStream.m (hunk 7 above)
Comment (thealienthing) [reply Applications/MIDIMonitor/SMMCombinationInputStream.m]: Sorry I missed this. I've never made a real PR until now.

Comment (krevis) [reply Applications/MIDIMonitor/SMMCombinationInputStream.m]: No worries! You have to start somewhere.

File: Applications/MIDIMonitor/SMMAppController.m (hunk 9 above)
Comment (krevis) [reply Applications/MIDIMonitor/SMMAppController.m]: OK, current window is probably fine; just wanted to make sure that whatever you were doing, you were doing it intentionally. 

It's possible that `[dc currentDocument]` is nil (for instance if no windows are open) so make sure the code does something sensible if that's the case.

File: Applications/MIDIMonitor/SMMAppController.h (hunk 3 above)
Comment (krevis) [reply Applications/MIDIMonitor/SMMAppController.h]: Right, leave the old one in place, not commented out, since you'll need it for migration.

File: Applications/MIDIMonitor/SMMAppController.m (hunk 9 above)
Comment (thealienthing) [reply Applications/MIDIMonitor/SMMAppController.m]: Good point. Nothing happens if there isn't a currently selected window. I'll see about making it select the most recently used window. What do you think?

Comment (krevis) [reply Applications/MIDIMonitor/SMMAppController.m]: Sounds good!  Whatever seems like the most reasonable user experience.

**Comment by thealienthing** [comment]
I've fixed all the issues you've mentioned and push them to my master branch. Is there anything I need to do on this pull request, or can you see those changes?

**Comment by krevis** [comment]
Got the rest of the changes, thanks! Will take a look over the holiday this week or the weekend.
